### PR TITLE
// Can I Haz a new Core ?

### PR DESCRIPTION
**Pull request by tchauviere**

### CODE REVIEW

**Comment by Quetzacoalt91** [comment]
```
WOW !
             Many commits !

   Such changes !!!!

            Awesome !
````
doge

File: Adapter/Adapter_EntityMapper.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+		if (!$should_cache_objects || !Cache::isStored($cache_id)) {
+			$sql = new DbQuery();
+			$sql->from($entity_defs['table'], 'a');
+			$sql->where('a.' . $entity_defs['primary'] . ' = ' . (int)$id);
+
+			// Get lang informations
+			if ($id_lang && isset($entity_defs['multilang']) && $entity_defs['multilang']) {
+				$sql->leftJoin($entity_defs['table'] . '_lang', 'b', 'a.' . $entity_defs['primary'] . ' = b.' . $entity_defs['primary'] . ' AND b.id_lang = ' . (int)$id_lang);
+				if ($id_shop && !empty($entity_defs['multilang_shop']))
+					$sql->where('b.id_shop = ' . $id_shop);
```
Comment (jeromenadaud) [inline Adapter/Adapter_EntityMapper.php]: Cast or escape please

File: Adapter/Adapter_EntityMapper.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	 * @throws PrestaShopDatabaseException
+	 */
+	public function load($id, $id_lang, $entity, $entity_defs, $id_shop, $should_cache_objects)
+	{
+		// Load object from database if object id is present
+		$cache_id = 'objectmodel_' . $entity_defs['classname'] . '_' . (int)$id . '_' . (int)$id_shop . '_' . (int)$id_lang;
+		if (!$should_cache_objects || !Cache::isStored($cache_id)) {
+			$sql = new DbQuery();
+			$sql->from($entity_defs['table'], 'a');
+			$sql->where('a.' . $entity_defs['primary'] . ' = ' . (int)$id);
```
Comment (jeromenadaud) [inline Adapter/Adapter_EntityMapper.php]: Should be escaped with back quotes

File: Adapter/Adapter_EntityMapper.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+			// Get shop informations
+			if (Shop::isTableAssociated($entity_defs['table']))
+				$sql->leftJoin($entity_defs['table'] . '_shop', 'c', 'a.' . $entity_defs['primary'] . ' = c.' . $entity_defs['primary'] . ' AND c.id_shop = ' . (int)$id_shop);
+			if ($object_datas = Db::getInstance()->getRow($sql)) {
+				if (!$id_lang && isset($entity_defs['multilang']) && $entity_defs['multilang']) {
+					$sql = 'SELECT * FROM `' . pSQL(_DB_PREFIX_ . $entity_defs['table']) . '_lang`
+								WHERE `' . bqSQL($entity_defs['primary']) . '` = ' . (int)$id
+						. (($id_shop && $entity->isLangMultishop()) ? ' AND `id_shop` = ' . $id_shop : '');
+					if ($object_datas_lang = Db::getInstance()->executeS($sql))
```
Comment (jeromenadaud) [inline Adapter/Adapter_EntityMapper.php]: cast or escape please

File: Adapter/Adapter_EntityMapper.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+								}
+							}
+				}
+				if ($should_cache_objects)
+					Cache::store($cache_id, $object_datas);
+			}
+		} else
+			$object_datas = Cache::retrieve($cache_id);
+
+		if ($object_datas) {
```
Comment (jeromenadaud) [inline Adapter/Adapter_EntityMapper.php]: Please move this assignations into a methode, in order to call it previously at line 71 and bypass Cache::retreive

File: Adapter/Adapter_EntityMetaDataRetriever.php [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        $metaData = new Core_Foundation_Database_EntityMetaData;
+
+        $metaData->setEntityClassName($className);
+
+        if (property_exists($className, 'definition')) {
+            // Legacy entity
+            $metaData->setTableName($className::$definition['table']);
+            $metaData->setPrimaryKeyFieldNames(array($className::$definition['primary']));
+        } else {
+            throw new Exception(
```
Comment (jeromenadaud) [inline Adapter/Adapter_EntityMetaDataRetriever.php]: Prefer to create and use AdapterException

File: Adapter/Adapter_ServiceLocator.php [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    private static $service_container;
+
+    public static function setServiceContainerInstance(Core_Foundation_IoC_Container $container)
+    {
+        self::$service_container = $container;
+    }
+
+    public static function get($serviceName)
+    {
+        return self::$service_container->make($serviceName);
```
Comment (jeromenadaud) [inline Adapter/Adapter_ServiceLocator.php]: Please check if self::$service_container is not empty and throw and exception if it is

File: Core/Business/CMS/Core_Business_CMS_CMSRepository.php [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+	private function getLanguageTableNameWithPrefix()
+	{
+		return $this->getTableNameWithPrefix() . '_lang';
+	}
+
+	public function i10nFindAll($id_lang, $id_shop)
+	{
+		$sql = '
+			SELECT *
+			FROM `'.$this->getTableNameWithPrefix().'` c
```
Comment (jeromenadaud) [inline Core/Business/CMS/Core_Business_CMS_CMSRepository.php]: Escape possible back quotes with bqSQL or other NEW function

File: Core/Business/CMS/Core_Business_CMS_CMSRepository.php [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+			JOIN `'.$this->getLanguageTableNameWithPrefix().'` cl ON c.`id_cms`= cl.`id_cms`
+			WHERE cl.`id_lang` = '.$id_lang.'
```
Comment (jeromenadaud) [inline Core/Business/CMS/Core_Business_CMS_CMSRepository.php]: cast or escape please

**Comment by xBorderie** [comment]
Great!

File: Adapter/Adapter_EntityMapper.php (hunk 1 above)
Comment (tchauviere) [inline Adapter/Adapter_EntityMapper.php]: :+1:

File: Adapter/Adapter_EntityMapper.php (hunk 2 above)
Comment (tchauviere) [inline Adapter/Adapter_EntityMapper.php]: :+1:

File: Adapter/Adapter_EntityMapper.php (hunk 3 above)
Comment (tchauviere) [inline Adapter/Adapter_EntityMapper.php]: :+1:

File: Adapter/Adapter_EntityMetaDataRetriever.php (hunk 5 above)
Comment (tchauviere) [inline Adapter/Adapter_EntityMetaDataRetriever.php]: :+1:

File: Adapter/Adapter_ServiceLocator.php (hunk 6 above)
Comment (tchauviere) [inline Adapter/Adapter_ServiceLocator.php]: :+1:

File: Core/Business/CMS/Core_Business_CMS_CMSRepository.php (hunk 7 above)
Comment (tchauviere) [inline Core/Business/CMS/Core_Business_CMS_CMSRepository.php]: :+1:

File: Core/Business/CMS/Core_Business_CMS_CMSRepository.php (hunk 8 above)
Comment (tchauviere) [inline Core/Business/CMS/Core_Business_CMS_CMSRepository.php]: :+1:

File: controllers/admin/AdminOrderPreferencesController.php [hunk 9]
```diff
@@ -169,7 +169,14 @@ public function __construct()
 		);
 
 		if (!Configuration::get('PS_ALLOW_MULTISHIPPING'))
+		{
 			unset($this->fields_options['general']['fields']['PS_ALLOW_MULTISHIPPING']);
+		}
```
Comment (jeromenadaud) [inline controllers/admin/AdminOrderPreferencesController.php]: norm please

File: controllers/front/OrderOpcController.php [hunk 10]
```diff
@@ -413,7 +413,12 @@ public function initContent()
 
 		$this->_processAddressFormat();
 
-		$this->setTemplate(_PS_THEME_DIR_.'order-opc.tpl');
+		if ((bool)Configuration::get('PS_ADVANCED_PAYMENT_API')) {
```
Comment (jeromenadaud) [inline controllers/front/OrderOpcController.php]: norm please

File: themes/default-bootstrap/order-carrier.tpl [hunk 11]
```diff
@@ -348,16 +348,22 @@
 				{/if}
 				{/if}
 			{/if}
-			{if $conditions AND $cms_id}
+			{if $conditions AND $cms_id AND !$advanced_payment_api}
```
Comment (jeromenadaud) [inline themes/default-bootstrap/order-carrier.tpl]: Please use &&

**Comment by jeromenadaud** [comment]
:pray:

**Comment by maximebiloe** [comment]
:muscle:

**Comment by tchauviere** [comment]
:tada:

**Comment by djfm** [comment]
Yes I think that's what the calculations do, but don't you mean rather this:

total tax amount on shipping is (60% of the shipping fee **x** 1,07 - 60% of the shipping fee) + (40% of the shipping fee **x** 1,19 - 40% of the shipping fee)

because your formula computes a negative value :)

To try it, install advancedeucompliance from [GitHub](https://github.com/PrestaShop/advancedeucompliance/tree/dev) (dev branch) and toggle the "use average tax" option.

The module needs the latest version of PrestaShop, so either the 1.6.1.0-RC3 package or the [1.6.1.0-RC3](https://github.com/PrestaShop/PrestaShop/tree/1.6.1.0-RC3) branch.

Then on the invoice the tax is "spread" over the actual tax rates so that you don't have a meaningless "11.80% tax" line but rather a shipping line at 7% and another shipping line at 19%.

**Pull request merged**